compiler/rustc_resolve/src/late.rs

```diff
@@ -1539,8 +1539,24 @@ impl<'a, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
                         .unwrap_or_else(|| self.fresh_binding(ident, pat.id, pat_src, bindings));
                     self.r.record_partial_res(pat.id, PartialRes::new(res));
                 }
-                PatKind::TupleStruct(ref path, ..) => {
-                    self.smart_resolve_path(pat.id, None, path, PathSource::TupleStruct(pat.span));
+                PatKind::TupleStruct(ref path, ref sub_patterns) => {
+                    let mut ident_bindings = vec![];
+                    for p in sub_patterns {
+                        if let PatKind::Ident(_, ref ident, None) = p.kind {
+                            ident_bindings.push(ident.clone())
+                        } else {
+                            // NOTE: Only handles patterns of the form
+                            // TupleStruct(a, b, ref c)
+                            ident_bindings.clear();
+                            break;
```

Comment:
It seems to me that instead of trying to do it only for idents you can store either the patterns themselves or their spans, right? Trying to destructure `Foo(Bar(Baz(())))` when the field in Foo is not visible should be an error.

Reply:
I can see if this is very difficult to add